### PR TITLE
Major Tags refactor

### PR DESCRIPTION
**Pull request by edgurgel**
This PR changes completely how Tags are defined. Now basic and custom tags use the same interface.

The main changes are:

* Tags now use the same behaviour that was added by @bluzky . Except that the `spec` function was changed to receive the `parser` so it can reference it if need be.
* Custom tags don't need to passed around as `options` for the `render` function. Soon we should not need to send the parser as option as well. 

The next version will completely break how custom tags work today but I think this is a much better way going forward as the parsing and rendering is completely customised giving much more power to the user.

I still want to send a PR with integration tests for each tag just like I did for `Assign` but it will be a separate PR.

@bluzky, would love to know your opinion about this PR

### CODE REVIEW

**Comment by bluzky** [comment]
Thanks @edgurgel, this is a really huge PR, I will review it carefully

**Comment by edgurgel** [comment]
@bluzky oh yeah no rush and thanks in advance!

**Review by bluzky — APPROVED** [review]
Hi @edgurgel, thanks for your awesome PR, I have chance to read all render logic 😄 . 
I really love your clever solution that using module name as tag and use it for render function instead of passing tag renderer via options.

I only have some minor comment for this.

File: lib/solid/parser/base.ex [hunk 1]
```diff
@@ -66,379 +35,40 @@ defmodule Solid.Parser.Base do
...
-
-      custom_tags =
-        if unquote(custom_tag_modules) != [] do
-          unquote(custom_tag_modules)
+        if custom_tag_modules != [] do
+          custom_tag_modules
           |> Enum.uniq()
-          |> Enum.reduce([], fn {tag_name, module}, acc ->
-            [tag(module.spec(), tag_name) | acc]
+          |> Enum.map(fn {tag_name, module} ->
```
[the remark this answers is not in the document]
Comment (edgurgel) [reply lib/solid/parser/base.ex:66]: Yes! It makes total sense

File: lib/solid/parser/base.ex [hunk 2]
```diff
@@ -1,12 +1,10 @@
...
-    custom_tag_names = custom_tags -- custom_tag_modules
+    custom_tag_modules = Keyword.get(opts, :custom_tags, [])
 
-    quote location: :keep do
+    quote location: :keep, bind_quoted: [custom_tag_modules: custom_tag_modules] do
       import NimbleParsec
-      alias Solid.Parser.{Literal, Variable, Argument}
+      alias Solid.Parser.{Literal, Variable, Argument, BaseTag}
 
       defp when_join(whens) do
```
[the remark this answers is not in the document]
Comment (edgurgel) [reply lib/solid/parser/base.ex:9]: Yup I completely forgot to move it! Nice catch!

File: lib/solid/parser/base.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (edgurgel) [reply lib/solid/parser/base.ex]: Done!

File: lib/solid/parser/base.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (edgurgel) [reply lib/solid/parser/base.ex]: Done!

**Pull request merged**